// sealious/cosealious/lib/mixins/resourceTypeCollection.js
"use strict";

Object.defineProperty(exports, "__esModule", {
	value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

exports.default = resourceTypeCollection;

var _react = require("react");

var _react2 = _interopRequireDefault(_react);

var _qwest = require("qwest");

var _qwest2 = _interopRequireDefault(_qwest);

var _merge = require("merge");

var _merge2 = _interopRequireDefault(_merge);

var _deepEqual = require("deep-equal");

var _deepEqual2 = _interopRequireDefault(_deepEqual);

var _clone = require("clone");

var _clone2 = _interopRequireDefault(_clone);

var _bluebird = require("bluebird");

var _bluebird2 = _interopRequireDefault(_bluebird);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var Loading = require("../loading");
var CachedHttp = require("../cached-http");

function resourceTypeCollection(ComponentClass) {
	var resourceTypeWrapper = function (_React$Component) {
		_inherits(resourceTypeWrapper, _React$Component);

		function resourceTypeWrapper(props) {
			_classCallCheck(this, resourceTypeWrapper);

			var _this = _possibleConstructorReturn(this, (resourceTypeWrapper.__proto__ || Object.getPrototypeOf(resourceTypeWrapper)).call(this, props));

			_this.state = {
				loading: false,
				resources: [],
				response: { attachments: {}, items: [] }
			};
			_this.generateQuery = _this.generateQuery.bind(_this);
			_this.fetch = _this.fetch.bind(_this);
			_this.refresh = _this.refresh.bind(_this);
			return _this;
		}

		_createClass(resourceTypeWrapper, [{
			key: "generateQuery",
			value: function generateQuery(props) {
				var query = {};
				if (props.filter) {
					query.filter = props.filter;
					for (var i in query.filter) {
						if (query.filter[i] == "undefined" || query.filter[i] === null) {
							delete query.filter;
						}
					}
				}
				var to_add = ["sort", "format", "search", "pagination"];
				to_add.forEach(function (key) {
					if (props[key]) {
						query[key] = props[key];
					}
				});
				return query;
			}
		}, {
			key: "reloadNeeded",
			value: function reloadNeeded(query) {
				return !(0, _deepEqual2.default)(query, this.state.last_query);
			}
		}, {
			key: "fetch",
			value: function fetch(query) {
				var _this2 = this;

				this.setState({
					loading: true
				});
				return CachedHttp.get(this.props.url, query, {
					cache: true
				}).then(function (response) {
					_this2.setState({
						loading: false,
						resources: _this2.props.customSort(response.items),
						response: response,
						last_query: (0, _clone2.default)(query)
					});
				});
			}
		}, {
			key: "refresh",
			value: function refresh(force) {
				var query = this.generateQuery(this.props);
				this.fetch(query);
			}
		}, {
			key: "componentDidMount",
			value: function componentDidMount() {
				this.refresh();
			}
		}, {
			key: "componentWillReceiveProps",
			value: function componentWillReceiveProps(next_props) {
				var _this3 = this;

				if (!this.props.will_not_update) setTimeout(function () {
					_this3.refresh();
				}, 0);
			}
		}, {
			key: "delete",
			value: function _delete(resource) {
				var _this4 = this;

				_qwest2.default.delete(this.props.url + "/" + resource.id, {}, { cache: true }).then(function () {
					_this4.refresh(true);
				});
			}
		}, {
			key: "render",
			value: function render() {
				if (this.state.loading) {
					return _react2.default.createElement(this.props.loadingComponent);
				}
				var customProps = _extends({}, this.state, {
					delete: this.delete
				});

				var child_props = (0, _merge2.default)(true, this.props, customProps);

				return _react2.default.createElement(ComponentClass, child_props);
			}
		}]);

		return resourceTypeWrapper;
	}(_react2.default.Component);

	resourceTypeWrapper.defaultProps = {
		pagination: undefined, // could be: {page: 1, items: 12}
		filter: {},
		format: {},
		search: "",
		url: "/api/v1/collections/users",
		loadingComponent: function loadingComponent() {
			return _react2.default.createElement(Loading);
		},
		customSort: function customSort(list) {
			return list;
		}
	};

	return resourceTypeWrapper;
}
